### PR TITLE
Create test case "IP Prefix Diversity": Connectivity04

### LINKED ISSUE #1102: Add Test Case checking for name server IPs prefix diversity
Currently we have a Test Case that checks for AS diversity ([Connectivity03](https://github.com/zonemaster/zonemaster/blob/master/docs/specifications/tests/Connectivity-TP/connectivity03.md)), i.e. that IP addresses of name servers of a domain are announced from different ASes. For those same IPs, and following the same general principle of network diversity but at a different level, the same could be done regarding the announced IP prefix they are tied to, i.e. that they are not announced in the same subnet.

Data used in Connectivity03 (Cymru/RIPE) already provide the necessary information. One question would be to whether extend this Test Case (update, rename it to something like "Network Diversity"), or rather create a new Test Case for this issue specifically.

### PR DESCRIPTION
**Pull request by tgreenx**
## Purpose

This PR proposes a specification for a new Test Case titled "IP Prefix Diversity". The proposed Test Case Identifier is Connectivity04.

## Context

Fixes #1102 , #1043 

## Changes

- New specification (`docs/public/specifications/tests/Connectivity-TP/connectivity04.md`)
- Some minor fixes in other documents

## How to test this PR

N/A. Documentation to review only.

### CODE REVIEW

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+   ("Name Server IP").
+
+2. For each name server IP in *Name Server IP* do:
+   1. Add IPv4 addresses to the *NS IPv4 IPs* set.
+   2. Add IPv6 addresses to the *NS IPv6 IPs* set.
+
+2. For each IP address ("NS IP Address") in *NS IPv4 IPs* and *NS IPv6 IPs*,
+   respectively, do:
+   1. Determine the IP prefix in which *NS IP Address* is announced using either
+      the *[Cymru Database]* or the *[RIPE Database]* (details in sections below).
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:85]: Shouldn't the selection be based on the value of *Prefix Database* from the input section?

**Review by marc-vanderwal — APPROVED** [review]
Looks good to me.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+6. Return the IP prefix.
+
+## Intercase dependencies
+
+None
+
+## Terminology
+
+* "Send" - The term is used when a DNS query is sent to
+  a specific name server (name server IP address).
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/public/specifications/tests/Connectivity-TP/connectivity04.md:202]: I think either solution could work.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+   ("Name Server IP").
+
+3. For each name server IP in *Name Server IP* do:
+   1. Add IPv4 addresses to the *NS IPv4 IPs* set.
+   2. Add IPv6 addresses to the *NS IPv6 IPs* set.
+
+4. For each IP address in *NS IPv4 IPs* and *NS IPv6 IPs* ("NS IP Address"),
+   respectively, do:
+   1. Determine the IP prefix in which *NS IP Address* is announced
+      using *Prefix Database* (details in sections below).
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:84]: I do not think it is obvious from this text that you should select either the "Cymru Prefix Lookup" section or "RIPE Prefix Lookup" section below. I will not be a problem for Zonemaster implementation but maybe for others that read the text.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+   2. Add found IP prefix, if any, with *NS IP Address* to the *IPv4 Prefix*
+      and *IPv6 Prefix* sets, respectively.
+
+5. If the *IPv4 Prefix* set is non-empty, then do:
+   1. For each IP prefix in the set that has two or more members, output
+      *[CN04_IPV4_SAME_PREFIX]* with the prefix and list of all members for that prefix.
+   2. For each IP prefix in the set that has exactly one member, output
+      *[CN04_IPV4_DIFFERENT_PREFIX]* with the combined set of their associated members.
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:92]: "For each IP prefix ... output [CN04_IPV4_DIFFERENT_PREFIX]" sounds like outputting the tag potentially several times. How about

```
   2. For all NS IP addresses that are unique for its IP prefix in *IPv4 Prefix* set
       output *[CN04_IPV4_DIFFERENT_PREFIX]* with the combined set of NS IP addresses.
```

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+6. If the *IPv6 Prefix* set is non-empty, then do:
+   1. For each IP prefix in the set that has two or more members, output
+      *[CN04_IPV6_SAME_PREFIX]* with the prefix and list of all members for that prefix.
+   2. For each IP prefix in the set that has exactly one member, output
+      *[CN04_IPV6_DIFFERENT_PREFIX]* with the combined set of their associated members.
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:98]: See IPv4 above.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+In other cases, no message or only messages with severity level
+*[INFO]* or *[NOTICE]*, the outcome of this Test Case is "pass".
+
+## Special procedural requirements
+
+This Test Case is dependent on one of two possible services that can provide
+ASN lookup (Cymru or RIPE RIS). The service must be available over the network.
+
+The *Child Zone* must be a valid name meeting
+"[Requirements and normalization of domain names in input]".
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:118]: The link is missing in the link section.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+### Cymru prefix lookup
+
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:121]: Should this section be a sub-section to "Special procedural requirements"? Maybe a header
```
## Prefix lookup methods

Use the prefix method set in *Prefix Database* in "Inputs".

```

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+5. [Send] *TXT Query* to the *Cymru Base Name*.
+
+6. If at least one of the following criteria is met, output
+   *[CN04_EMPTY_PREFIX_SET]* and exit this lookup:
+   1. The [DNS Response] has the [RCODE Name] NXDomain.
+   2. The [DNS Response] has the [RCODE Name] NoError and an empty answer section.
+
+7. If at least one of the following criteria is met, output
+   *[CN04_ERROR_PREFIX_DATABASE]* and exit this lookup:
+   1. There is no response.
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:152]: "no DNS response"

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+   2. The [DNS Response] does not have the [RCODE Name] NoError.
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:153]: It should rather be "NoError or NXDomain".

**Review by matsduf — CHANGES_REQUESTED** [review]

**Comment by matsduf** [comment]
To be consistent with e.g. #1150 then this specification should also include something about concatenation of strings in a TXT record RDATA.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md (hunk 7 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Connectivity-TP/connectivity04.md:127]: Don’t forget to update the table of contents too. 🙂

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md (hunk 9 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Connectivity-TP/connectivity04.md:159]: True, but repeating “NoError or NXDomain” would make the intent clearer.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+This Test is done separately on IPv4 and IPv6, and both must match the criterion.
+
+[RFC 2182, section 3.1][RFC 2182#3.1], clearly specifies that distinct authoritative
+name servers for a child domain should be placed in different topological and 
+geographical locations. The objective is to minimise the likelihood of a single 
+failure disabling all of them.
+
+## Scope
+
+It is assumed that *Child Zone* is also tested and reported by [Connectivity01].
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:35]: Link for `[Connectivity01]` is missing.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+   1. Determine the IP prefix in which *NS IP Address* is announced
+      using *Prefix Database*. See [Prefix Lookup Methods] section below.
+   2. Add found IP prefix, if any, with *NS IP Address* to the *IPv4 Prefix*
+      and *IPv6 Prefix* sets, respectively.
+
+5. If the *IPv4 Prefix* set is non-empty, then do:
+   1. For each IP prefix in the set that has two or more members, output
+      *[CN04_IPV4_SAME_PREFIX]* with the prefix and list of all members for that prefix.
+   2. For all IP prefixes in the set that have exactly one member, output
+      *[CN04_IPV4_DIFFERENT_PREFIX]* with the combined set of their associated members.
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:93]: "with the combined set of the name server IP addresses" makes it easier to read.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+
+6. If the *IPv6 Prefix* set is non-empty, then do:
+   1. For each IP prefix in the set that has two or more members, output
+      *[CN04_IPV6_SAME_PREFIX]* with the prefix and list of all members for that prefix.
+   2. For all IP prefixes in the set that have exactly one member, output
+      *[CN04_IPV6_DIFFERENT_PREFIX]* with the combined set of their associated members.
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:99]: Same as IPv4 above.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+* [Objective](#objective)
+* [Scope](#scope)
+* [Inputs](#inputs)
+* [Summary](#summary)
+* [Test procedure](#test-procedure)
+* [Outcome(s)](#outcomes)
+* [Special procedural requirements](#special-procedural-requirements)
+* [Prefix lookup methods](#prefix-lookup-methods)
+* [Cymru prefix lookup](#cymru-prefix-lookup)
+* [RIPE prefix lookup](#ripe-prefix-lookup)
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:17]: I think these two should be indented because they are sub-headers.

**Review by matsduf — CHANGES_REQUESTED** [review]

**Comment by matsduf** [comment]
This TC does not specify that CNAME should be followed, which Zonemaster by default does not do. I do not expect any CNAME records from Cymru, but just to be sure.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+[CN04_IPV6_DIFFERENT_PREFIX]:                                   #outcomes
+[CN04_IPV6_SAME_PREFIX]:                                        #outcomes
+[Concatenate]:                                                  #terminology
+[Connectivity01]:                                               ../Connectivity-TP/connectivity01.md
+[CRITICAL]:                                                     ../SeverityLevelDefinitions.md#critical
+[Cymru Database]:                                               #cymru-prefix-lookup
+[DEBUG]:                                                        ../SeverityLevelDefinitions.md#notice
+[DNS Query and Response Defaults]:                              ../DNSQueryAndResponseDefaults.md
+[DNS Query]:                                                    ../DNSQueryAndResponseDefaults.md#default-setting-in-dns-query
+[DNS Response]:                                                 ./DNSQueryAndResponseDefaults.md#default-handling-of-a-dns-response
```
Comment (marc-vanderwal) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:232]: This link is broken (probably due to a typo: `.` where `..` was meant).

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+
+[Argument List]:                                                https://github.com/zonemaster/zonemaster-engine/blob/master/docs/logentry_args.md
+[CN04_EMPTY_PREFIX_SET]:                                        #outcomes
+[CN04_ERROR_PREFIX_DATABASE]:                                   #outcomes
+[CN04_IPV4_DIFFERENT_PREFIX]:                                   #outcomes
+[CN04_IPV4_SAME_PREFIX]:                                        #outcomes
+[CN04_IPV6_DIFFERENT_PREFIX]:                                   #outcomes
+[CN04_IPV6_SAME_PREFIX]:                                        #outcomes
+[Concatenate]:                                                  #terminology
+[Connectivity01]:                                               ../Connectivity-TP/connectivity01.md
```
Comment (marc-vanderwal) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:226]: Can’t this link target be simplified to just `connectivity01.md`?

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+[CRITICAL]:                                                     ../SeverityLevelDefinitions.md#critical
+[Cymru Database]:                                               #cymru-prefix-lookup
+[DEBUG]:                                                        ../SeverityLevelDefinitions.md#notice
+[DNS Query and Response Defaults]:                              ../DNSQueryAndResponseDefaults.md
+[DNS Query]:                                                    ../DNSQueryAndResponseDefaults.md#default-setting-in-dns-query
+[DNS Response]:                                                 ./DNSQueryAndResponseDefaults.md#default-handling-of-a-dns-response
+[ERROR]:                                                        ../SeverityLevelDefinitions.md#error
+[INFO]:                                                         ../SeverityLevelDefinitions.md#info
+[IP to ASN Mapping]:                                            https://team-cymru.com/community-services/ip-asn-mapping/#dns
+[Message Tag Specification]:                                    ../../../../internal/templates/specifications/tests/MessageTagSpecification.md
```
Comment (marc-vanderwal) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:236]: That link is not used anywhere in the document.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+[RFC 1035#3.5]:                                                 https://datatracker.ietf.org/doc/html/rfc1035#section-3.5
+[RFC 2182#3.1]:                                                 https://datatracker.ietf.org/doc/html/rfc2182#section-3.1
+[RFC 3596#2.5]:                                                 https://datatracker.ietf.org/doc/html/rfc3596#section-2.5
+[RFC7208#3.3]:                                                  https://datatracker.ietf.org/doc/html/rfc7208#section-3.3
+[RIPE Database]:                                                #ripe-prefix-lookup
+[RISwhois]:                                                     https://www.ripe.net/analyse/archived-projects/ris-tools-web-interfaces/riswhois
+[Send]:                                                         #terminology
+[Severity Level Definitions]:                                   ../SeverityLevelDefinitions.md
+[Test Case Identifier Specification]:                           ../../../../internal/templates/specifications/tests/TestCaseIdentifierSpecification.md
```
Comment (marc-vanderwal) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:252]: Unused link.

**Review by marc-vanderwal — CHANGES_REQUESTED** [review]
One link is still broken; a few more are unused and can be removed. Otherwise, it’s looking fine to me.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md (hunk 16 above)
Comment (matsduf) [reply docs/public/specifications/tests/Connectivity-TP/connectivity04.md:236]: For template only.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md (hunk 17 above)
Comment (matsduf) [reply docs/public/specifications/tests/Connectivity-TP/connectivity04.md:252]: For template only.

**Review by matsduf — COMMENTED** [review]
I agree with Marc's comments, else I can approve.

**Review by matsduf — DISMISSED** [review]

**Review by marc-vanderwal — APPROVED** [review]
Looks good to me.

**Comment by matsduf** [comment]
Maybe "network prefix" is a better term. See https://www.ripe.net/about-us/press-centre/understanding-ip-addressing and https://en.wikipedia.org/wiki/IP_address#Subnetworks

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 18]
```diff
@@ -0,0 +1,252 @@
+# CONNECTIVITY04: IP Prefix Diversity
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:1]: @tgreenx, maybe "network prefix" is a better term. See https://www.ripe.net/about-us/press-centre/understanding-ip-addressing and https://en.wikipedia.org/wiki/IP_address#Subnetworks

Such a change would have consequences in many places.

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+## Summary
+
+Message Tag                 | Level   | Arguments                   | Message ID for message tag
+:---------------------------|:--------|:----------------------------|:------------------------------------------------------------------------------------------------
+CN04_EMPTY_PREFIX_SET       | ERROR   | ns_ip                       | Prefix database returned no information for IP address {ns_ip}.
+CN04_ERROR_PREFIX_DATABASE  | ERROR   | ns_ip                       | Prefix database error. No data to analyze for IP address {ns_ip}.
+CN04_IPV4_SAME_PREFIX       | WARNING | ns_ip_list, ip_prefix       | The following name server(s) are announced in the same IPv4 prefix ({ip_prefix}): "{ns_ip_list}"
+CN04_IPV4_DIFFERENT_PREFIX  | INFO    | ns_ip_list                  | The following name server(s) are announced in unique IPv4 prefix(es): "{ns_ip_list}"
+CN04_IPV6_SAME_PREFIX       | WARNING | ns_ip_list, ip_prefix       | The following name server(s) are announced in the same IPv6 prefix ({ip_prefix}): "{ns_ip_list}"
+CN04_IPV6_DIFFERENT_PREFIX  | INFO    | ns_ip_list                  | The following name server(s) are announced in unique IPv6 prefix(es): "{ns_ip_list}"
```
Comment (matsduf) [inline docs/public/specifications/tests/Connectivity-TP/connectivity04.md:58]: `ns_list` would be an improvement compared to `ns_ip_list`

File: docs/public/specifications/tests/Connectivity-TP/connectivity04.md (hunk 18 above)
Comment (marc-vanderwal) [reply docs/public/specifications/tests/Connectivity-TP/connectivity04.md:1]: I don’t know. I don’t have any issue with either “IP prefix” or “network prefix”. However, if it turns out that “IP prefix” is a bad term, I suggest opening a separate issue.

**Review by marc-vanderwal — APPROVED** [review]
It’s indeed better to list the name server names, rather than their IPs, in the messages. Looks good to me.

**Pull request merged**